src/inmanta/references.py

```diff
@@ -169,6 +169,33 @@ def get_arg_value(
         return resource.get_reference_value(self.id, logger)
 
 
+class MutatedJsonArgument(Argument):
+    """
+    Json-like argument to a reference that contains reference in the json structure
+
+    It is stored as
+    1. a json value with all reference replaced by None
+    2. a mapping of dictpaths to references. Each dict path indicates where the reference should be inserted
+
+    """
+
+    type: typing.Literal["mjson"] = "mjson"
+    value: StrictJson
+    references: dict[str, ReferenceArgument]
+
+    def get_arg_value(
+        self,
+        resource: "inmanta.resources.Resource",
+        logger: "handler.LoggerABC",
+    ) -> object:
+        start_value: JsonType = self.value
+        for destination, valueref in self.references.items():
+            value = valueref.get_arg_value(resource, logger)
+            dict_path_expr = dict_path.to_path(destination)
+            dict_path_expr.set_element(start_value, value)
```

Comment:
For list indices, this requires the improved dict path, doesn't it?

Reply:
I'm going to pick that up next